### ISSUE TITLE
RefResolver is not working with local refs in 4.18 alpha

### ISSUE THREAD

**Issue by robherring** [python-jsonschema/jsonschema#1085]
4.18 alpha is not working for dtschema where we have a custom RefResolver instance (for a custom http handler to do local file lookups and load yaml files). I initially thought the problem was more complex chain of $ref's, but in coming up with a simplified example, it's just local refs not working. I've reproduced this with the jsonschema cli by forcing the RefResolver to be used by passing --base-uri like this:

echo '{}' | jsonschema --base-uri "" ~/.local/lib/python3.11/site-packages/jsonschema_specifications/schemas/draft201909/metaschema.json

Running this on v4.17.3 passes. 

Note that draft7 schema works because no $ref is reached.

**Comment by robherring** [comment]
Note that setting base-uri to "https://json-schema.org/draft/2019-09/schema" works in the above case. However, an empty base matches what the RefResolver in dtschema uses:

resolver = jsonschema.RefResolver('', None, handlers=handlers)

This is because the exact schema used varies and the RefResolver is instantiated before the schema is known. That's at least partly because there's still cases of iter_error(schema) being used.

**Comment by Julian** [comment]
Do you have an example that doesn't involve `$dynamicRef` / `$recursiveRef`, which weren't really supported previously in 4.17? (Supporting them was part of the reason for the rewrite in general) That seems more related to the error there than local refs. E.g. replacing your example with `draft7` seems to work fine (where those metaschemas do not use the above of course).

(Also, unrelated, but it is not part of `jsonschema-specifications'` public API to pull its internal data files out from within it, but I'm assuming that's just for examples' sake. But its public API is the one that's documented using `jsonschema_specifications.REGISTRY`)

**Comment by robherring** [comment]
> Do you have an example that doesn't involve `$dynamicRef` / `$recursiveRef`, which weren't really supported previously in 4.17? (Supporting them was part of the reason for the rewrite in general) That seems more related to the error there than local refs. E.g. replacing your example with `draft7` seems to work fine (where those metaschemas do not use the above of course).

As I noted draft7 happens to work because it doesn't have a top-level $ref. So you need make it descend a path that does. Like this:

echo '{"maxLength": "foo"}' | jsonschema --base-uri "" ~/.local/lib/python3.11/site-packages/jsonschema_specifications/schemas/draft7/metaschema.json

> 
> (Also, unrelated, but it is not part of `jsonschema-specifications'` public API to pull its internal data files out from within it, but I'm assuming that's just for examples' sake. But its public API is the one that's documented using `jsonschema_specifications.REGISTRY`)

Yep, that's just for convenience of this example.

**Comment by jsolbrig** [comment]
I don't know if @robherring is using `unresolvedProperties: false`, but I'm running into a similar error when combining `unresolvedProperties: false` with references.

To demonstrate, if I attempt to validate the following document:
```json
{
    "foo": "bar"
}
```

The following pair of schema work as expected:
```json
{       
    "$id": "base",
    "$ref": "referenced"                                                                                                                                                                                                                                                                                        
}
```
```json
{                                                                                                                                                                                                                                                                                                                     
    "$id": "referenced",
    "type": "object",
    "properties": {
        "foo": {
            "type": "string"
        }
    }   
}
```

If, however, I add `unevaluatedProperties: false` to `base` like this:
```yaml
{       
    "$id": "base",
    "unevaluatedProperties": false,                                                                                                                                                                                                                                                                                   
    "$ref": "referenced"
}  
```
then validate the same json, I get the following error:
```
Traceback (most recent call last):
  File "./test.py", line 22, in <module>
    validators["base"].validate(schema)
  File ".../jsonschema/validators.py", line 393, in validate
    for error in self.iter_errors(*args, **kwargs):
  File ".../jsonschema/validators.py", line 330, in iter_errors
    for error in errors:
  File ".../jsonschema/_validators.py", line 411, in unevaluatedProperties
    evaluated_keys = find_evaluated_property_keys_by_schema(
  File ".../jsonschema/_utils.py", line 262, in find_evaluated_property_keys_by_schema
    resolved = validator._resolver.lookup(schema["$ref"])
  File ".../referencing/_core.py", line 582, in lookup
    raise exceptions.Unresolvable(ref=ref) from None
referencing.exceptions.Unresolvable: referenced
```

**Comment by jsolbrig** [comment]
@Julian yeah, I see what you mean. I also discovered that the above problem was my problem anyway. I shouldn't be writing code too late at night I guess.

In order to keep things clean, would you prefer if I just delete my comment here since it's not actually a bug?

**Comment by Julian** [comment]
I'll hide it, but all good!

**Comment by robherring** [comment]
This is still failing for me. The above test works, but if I change the test case to Draft201909 it fails:

```
    def test_pointer_within_schema_with_different_id(self):
        """
        See #1085.
        """
        schema = validators.Draft201909Validator.META_SCHEMA
        resolver = validators._RefResolver("", schema)
        validator = validators.Draft201909Validator(schema, resolver=resolver)
        self.assertFalse(validator.is_valid({"maxLength": "foo"}))

```

Which results in:

```
______________________ TestRefResolver.test_pointer_within_schema_with_different_id _______________________
                                                                                                           
self = <jsonschema.validators._RefResolver object at 0xfffed6a527d0>, url = 'meta/core'                    
                                                                                                           
    def resolve_from_url(self, url):                                                                       
        """                                                                                                
        Resolve the given URL.                                                                             
        """                                                                                                
        url, fragment = urldefrag(url)                                                                     
        if not url:                                                                                        
            url = self.base_uri                      
                                                                                                           
        try:                                                                                               
>           document = self.store[url]                                                                     
                                                                                                           
jsonschema/validators.py:1078:                                                                             
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
                                                                                                           
self = {'https://json-schema.org/draft/2019-09/meta/content': {'$schema': 'https://json-schema.org/draft/20
19-09/schema', '$i...bject', 'additionalProperties': {'anyOf': [{'$recursiveRef': '#'}, {'$ref': 'meta/vali
dation#/$defs/stringArray'}]}}}}}                                                                          
uri = 'meta/core'                                                                                          
                                                                                                           
    def __getitem__(self, uri):                                                                            
>       return self.store[self.normalize(uri)]       
E       KeyError: 'meta/core'                                                                              
                                                                                                           
jsonschema/_utils.py:20: KeyError                                                                                                                                                                                     
During handling of the above exception, another exception occurred:                                        
                                                                                                           
self = <jsonschema.validators._RefResolver object at 0xfffed6a527d0>, url = 'meta/core'

    def resolve_from_url(self, url):
        """
        Resolve the given URL.
        """
        url, fragment = urldefrag(url)
        if not url:
            url = self.base_uri
     
        try:
            document = self.store[url]
        except KeyError:
            try:
>               document = self.resolve_remote(url)

```

**Comment by Julian** [comment]
Can you provide a still-failing example that doesn't involve `recursiveRef`?

**Comment by robherring** [comment]
This does it:
```

diff --git a/jsonschema/tests/test_validators.py b/jsonschema/tests/test_validators.py
index 00e0c403294f..a0e508c45ba5 100644
--- a/jsonschema/tests/test_validators.py
+++ b/jsonschema/tests/test_validators.py
@@ -2333,6 +2333,7 @@ class TestRefResolver(TestCase):
         See #1085.
         """
         schema = validators.Draft7Validator.META_SCHEMA
+        schema = {"allOf": [{"$ref": schema['$id']}]}
         resolver = validators._RefResolver("", schema)
         validator = validators.Draft7Validator(schema, resolver=resolver)
         self.assertFalse(validator.is_valid({"maxLength": "foo"}))

```

**Comment by Julian** [comment]
Great, thanks!

**Comment by Julian** [comment]
The second example should now also be addressed.

**Comment by robherring** [comment]
Unfortunately, it's still failing for me. I'll try to come up with another example.

**Comment by robherring** [comment]
This still fails. Note that referrer is None in this case:

```
        another = {
            "$defs": {
                "ref": {
                    "$ref": validators.Draft7Validator.META_SCHEMA["$id"],
                },
            },
            "allOf": [{"$ref": "#/$defs/ref"}],
        }
        two = validators._RefResolver("", None)
        validator = validators.Draft7Validator(another, resolver=two)
        self.assertFalse(validator.is_valid({"maxLength": "foo"}))

```

**Comment by robherring** [comment]
It was valid in 4.17 in that it all worked fine. I used none to instantiate the resolver once instead of ~3K times for each schema. I could rework it, but that wouldn't fix existing releases. Also, if I'm making changes, I might as well use referencing as I already have that working (minus the performance issue).

**Comment by Julian** [comment]
That's not what valid means, there's no promise of backwards compatibility for things that have behavior X but are against the docs, have no tests, aren't part of the public API, etc.

I don't see what this has to do with your issue, which might happen just as well if you *do* provide the correct type of object there (a schema, perhaps `{}`), as doing so probably has nothing to do with the performance optimization you're talking about -- though your reference doesn't look correct in that case, as a pointer `#/foo/bar` needs to know the schema it's sitting within, so I'm not sure what did or didn't work but it'd seem pretty unlikely whatever it was was correctly working according to the spec.

In short, you need to be a bit better at filing bugs, it's too much effort on my part to pull detail out of you -- it's on you the bug reporter to come up with an example, and to come up with one that's not against the spec! The same applies to #1088 where you haven't provided any information, so there's nothing to look at there.

**Comment by robherring** [comment]
> That's not what valid means, there's no promise of backwards compatibility for things that have behavior X but are against the docs, have no tests, aren't part of the public API, etc.
> 
> I don't see what this has to do with your issue, which might happen just as well if you _do_ provide the correct type of object there (a schema, perhaps `{}`), as doing so probably has nothing to do with the performance optimization you're talking about -- though your reference doesn't look correct in that case, as a pointer `#/foo/bar` needs to know the schema it's sitting within, so I'm not sure what did or didn't work but it'd seem pretty unlikely whatever it was was correctly working according to the spec.

Presumably 4.17 just got the Validator's schema if referrer was empty? Maybe it has nothing to do with what referrer is. I'll try some more experiments. It seems to me that #1092 could be related. It's some combination of local and external refs.

> In short, you need to be a bit better at filing bugs, it's too much effort on my part to pull detail out of you -- it's on you the bug reporter to come up with an example, and to come up with one that's not against the spec! The same applies to #1088 where you haven't provided any information, so there's nothing to look at there.

I debugged it and explained exactly what is happening, and provided a partial fix. I was really just looking for some thoughts and guidance on a better fix.
